tests/validation/tests/v3_api/common.py

```diff
@@ -2735,3 +2735,111 @@ def get_node_details(cluster, client):
         if node.worker:
             break
     return client, node
+
+
+def test_reader(file_path=None):
+    """
+    This method generates test cases from an input file and return the result
+    that can be used to parametrize pytest cases
+    :param file_path: the path to the JSON file for test cases
+    :return: a list of tuples of (cluster_role, command, authorization)
+    """
+    if file_path is None:
+        pytest.fail("no file is provided")
+    with open(file_path) as reader:
+        test_cases = json.loads(reader.read().replace("{resource_root}",
+                                                      DATA_SUBDIR))
+    output = []
+    for cluster_role, checks in test_cases.items():
+        for check in checks:
+            output.append((cluster_role,
+                           check['command'],
+                           check['authorization']))
+    return output
```

Comment:
In general for structuring the test cases input, it might be easier to have something with the following format:
```
roles: [
    {
        name: "monitoring-admin" # role name
        apply_before: [] # any test cases to apply with cluster-admin permissions first in case resources should already exist before other test cases are run
        should_pass: ["create-configmap", "get-configmap", ...] # test cases that need to pass
        should_fail: [] # test cases that should fail
    },
    {
        name: "monitoring-edit"
        apply_before: []
        should_pass: ["create-configmap", "get-configmap", ...]
        should_fail: []
    },
    {
        # What we expect from monitoring-view before an admin creates resources in a cluster
        name: "monitoring-view"
        apply_before: []
        should_pass: []
        should_fail: ["create-configmap", ...]
    },
    {
        # What we expect from monitoring-view after an admin creates resources in a cluster
        name: "monitoring-view"
        apply_before: ["create-configmap", ...]
        should_pass: ["get-configmap", ...]
        should_fail: []
    }
],
test_cases: {
    "create-configmap": {
        command: "create configmap my-config --from-literal=key1=config1"
    },
    "get-configmap": {
        command: "get configmap"
    },
    ....
}
```

Within `test_reader`, you can then convert each role entry into the corresponding:
- `cluster_role` (`roles[i].name`)
- `command` (`test_cases[roles[i].should_pass[j]` or `test_cases[roles[i].should_fail[j]`)
- `authorization` (whether you are iterating through test cases in `should_pass` or `should_fail`)
- `test_state` (list of `test_cases[roles[i].apply_before[j]]`)

The current structure ends up repeating the same underlying test cases with just a different `authorization` flag, which makes it harder to catch if minor changes will be necessary in the future. I'm also not sure how you would be able to test getting a specific resource (which tests the GET permissions as opposed to LIST permissions) without having some initial cluster state that needs to exist for view permissions.